rp2040-hal/src/gpio.rs

```diff
@@ -1,132 +1,337 @@
+// const FUNCTION_PIO1: u8 = 7;
+// const FUNCTION_CLOCK: u8 = 8;
+// const FUNCTION_USB: u8 = 9;
 
 macro_rules! gpio {
-    ($GPIOX:ident, $gpiox:ident, $PADSX:ident, [
-        $($PXi:ident: ($pxi:ident, $i:expr),)+
+    ($GPIOX:ident, $gpiox:ident, $PADSX:ident, $padsx:ident, $gpioxs:expr, [
+        $($PXi:ident: ($pxi:ident, $i:expr, $is:expr),)+
     ]) => {
-        impl GpioExt for pac::$GPIOX {
-            type Parts = Parts;
-
-            fn split(self, resets: &mut pac::RESETS) -> Parts {
-                resets.reset.modify(|_, w| w.$gpiox().clear_bit());
-                Parts {
-                    $(
-                        $pxi: $PXi { _mode: PhantomData },
-                    )+
+        #[doc = "HAL objects for the "]
+        #[doc = $gpioxs]
+        #[doc = " bank of GPIO pins"]
+        pub mod $gpiox {
+            use core::convert::Infallible;
+            use core::marker::PhantomData;
+            use embedded_hal::digital::v2::{InputPin, OutputPin, StatefulOutputPin};
+            use super::*;
+
+            impl GpioExt<pac::$PADSX, pac::SIO> for pac::$GPIOX {
+                type Parts = Parts;
+
+                fn split(self, pads: pac::$PADSX, sio: pac::SIO, resets: &mut pac::RESETS) -> Parts {
+                    resets.reset.modify(|_, w| w.$gpiox().clear_bit());
+                    Parts {
+                        _pads: pads,
+                        _sio: sio,
+                        $(
+                            $pxi: $PXi { _mode: PhantomData },
+                        )+
+                    }
                 }
             }
-        }
 
-        pub struct Parts {
-            $(
-                pub $pxi: $PXi<Input<Floating>>,
-            )+
-        }
+            #[doc = "Struct containing HAL objects for all the "]
+            #[doc = $gpioxs]
+            #[doc = " pins"]
+            pub struct Parts {
+                _pads: pac::$PADSX,
+                _sio: pac::SIO,
+                $(
+                    #[doc = "GPIO pin "]
+                    #[doc = $is]
+                    pub $pxi: $PXi<Unknown>,
+                )+
+            }
+
+            // Puts pad in same state as post-reset, according to datasheet
+            fn setup_pad_io(pads: &pac::$padsx::RegisterBlock, index: usize) {
+                pads.gpio[index].modify(|_, w| w.ie().set_bit()
+                                                .od().clear_bit()
+                                                .pue().clear_bit()
+                                                .pde().set_bit()
+                                                .schmitt().set_bit()
+                                                .drive()._4m_a()
+                                                .slewfast().clear_bit());
+            }
 
-        $(
-            pub struct $PXi<MODE> {
-                _mode: PhantomData<MODE>,
+            fn set_gpio_function(gpios: &pac::$gpiox::RegisterBlock, index: usize, function: u8) {
```

Comment:
Maybe we could turn the function to an `enum` instead of a `u8`. Otherwise it gets a bit harder to explain why the `unsafe` block is actually safe.